### PR TITLE
Copter: Stand-By mode for use with multiple flight controllers running in parallel

### PR DESCRIPTION
**Pull request by lthall**
StandBy is the ability to have multiple flight controllers working together where only one of these is the controller in command (CIC). The "ride along" controllers can experience significant I-term build up as they may have different attitude solution compared to the CIC. The secondary controllers may also be placed in a more basic mode. For example, during an auto mission the secondary flight controller may be held in Loiter. 

The functionality in this PR is designed to be a minimum impact standby functionality for ArduCopter.

The requirements of this capability are:
- Remove I term build up
- Remove integrated build up sources
- Not trigger crash checks, parachute, or landing detectors.
- Ensure the aircraft will still fly effectively in Stabilize, Alt Hold, Loiter and continue to stabilize the aircraft in other modes. This is to ensure the aircraft will continue to be controllable in the event that a secondary flight controller is made CIC without disabling the Standby state.

This will be the first introductory implementation of this functionality to facilitate development in this area.

### CODE REVIEW

**Comment by proficnc** [comment]
Love it! 
Thank you Leonard! This is absolutely awesome!

**Comment by lukezhqin** [comment]
good! SO you guys want to add a redundand CPU architecture?

**Comment by amilcarlucas** [comment]
I guess this is more a redundant CUBE architecture. I like it. A lot !!!

**Comment by proficnc** [comment]
Not just Cube, but yes, multiple Autopilots allowing for true redundancy.

Leonard has done some awesome work in this area

**Comment by lthall** [comment]
@amilcarlucas There is no aspect of this that is hardware dependent. It will enable a variety of methods and approaches to be further developed. This patch provides the most basic functionality required to have an autopilot in an armed and flying state while not in control.

File: ArduCopter/Copter.h [hunk 1]
```diff
@@ -766,6 +768,9 @@ class Copter : public AP_HAL::HAL::Callbacks {
     // landing_gear.cpp
     void landinggear_update();
 
+    // stand_by.cpp
+    void standby_update();
```
Comment (rmackay9) [inline ArduCopter/Copter.h]: at the risk of being annoying, maybe we can consolidate the naming to be either "standby" or "stand_by" including the naming of the .cpp file?  I personally prefer "standby" without the underscore but I don't mind too much either way.  the justification for this is just that it will make searches easier.

File: ArduCopter/crash_check.cpp [hunk 2]
```diff
@@ -23,6 +23,12 @@ void Copter::crash_check()
         return;
     }
 
+    // exit immediately if in standby
+    if (copter.standby_active == true) {
```
Comment (rmackay9) [inline ArduCopter/crash_check.cpp]: we can remove the "copter." and "== true" parts from this "if" and a few lower down because standby_active is already a boolean and this method is in the Copter class.

File: ArduCopter/land_detector.cpp [hunk 3]
```diff
@@ -54,7 +54,8 @@ void Copter::update_land_detector()
 #endif
             set_land_complete(false);
         }
-    } else {
+    } else if (copter.standby_active == false) {
```
Comment (rmackay9) [inline ArduCopter/land_detector.cpp]: we can make this one !stsandby_active (i.e. remove the "== false" part).  also no need for the .copter part.

File: ArduCopter/Copter.cpp [hunk 4]
```diff
@@ -144,6 +144,7 @@ const AP_Scheduler::Task Copter::scheduler_tasks[] = {
     SCHED_TASK(ekf_check,             10,     75),
     SCHED_TASK(gpsglitch_check,       10,     50),
     SCHED_TASK(landinggear_update,    10,     75),
+    SCHED_TASK(standby_update,        100,     75),
```
Comment (rmackay9) [inline ArduCopter/Copter.cpp]: nitpick: there's an extra space before the "75"

File: ArduCopter/stand_by.cpp [hunk 5]
```diff
@@ -0,0 +1,22 @@
+#include "Copter.h"
+
+
+// Run standby functions at 100Hz
```
Comment (rmackay9) [inline ArduCopter/stand_by.cpp]: really nice to see this explanation here!

File: ArduCopter/land_detector.cpp (hunk 3 above)
Comment (rmackay9) [reply ArduCopter/land_detector.cpp]: one thing to be careful about here is that land_detector_maybe could get stuck as true because the land_detector_count variable will not be being updated while standby_active.  I suspect it's OK.. it will probably clear soon after the standby_active becomes false but worth thinking about.

Another option might be to change the cascading ifs a bit to add a few lines like this:

```
} else if (standby_active) {
    // land detector will not run in standby mode
    land_detector_count = 0;
} else {
    // all the normal stuff
```

File: libraries/AC_AttitudeControl/AC_PosControl.cpp [hunk 6]
```diff
@@ -763,6 +763,23 @@ void AC_PosControl::init_xy_controller()
     init_ekf_xy_reset();
 }
 
+/// standby_xyz_reset - resets I terms and removes position error
+///     This function will let Loiter and Alt Hold continue to operate
+///     in the event that the flight controller is in control of the
```
Comment (rmackay9) [inline libraries/AC_AttitudeControl/AC_PosControl.cpp]: should there be a "not" in this sentence?

File: libraries/AC_AttitudeControl/AC_PosControl.cpp [hunk 7]
```diff
@@ -763,6 +763,23 @@ void AC_PosControl::init_xy_controller()
...
+///     aircraft when in standby.
+void AC_PosControl::standby_xyz_reset()
+{
+    _pid_accel_z.set_integrator(0.0f);
+
+    _pos_target = _inav.get_position();
+
+    // initialise I terms from lean angles
```
Comment (rmackay9) [inline libraries/AC_AttitudeControl/AC_PosControl.cpp]: is this an out-of-date comment or is it missing the call to "lean_angles_to_accel"?

File: libraries/AC_WPNav/AC_Loiter.cpp [hunk 8]
```diff
@@ -96,7 +96,7 @@ void AC_Loiter::init_target(const Vector3f& position)
 
     // initialise desired acceleration and angles to zero to remain on station
     _predicted_accel.zero();
-    _desired_accel = _predicted_accel;
```
Comment (rmackay9) [inline libraries/AC_WPNav/AC_Loiter.cpp]: these two changes in AC_Loiter look like they're not triggered from standby mode so they will always apply.  Are these unrelated changes?  .. and the remove of the set_max_speed_xy() and set_max_accel_xy() is intentional?

File: ArduCopter/stand_by.cpp (hunk 5 above)
Comment (peterbarker) [inline ArduCopter/stand_by.cpp]: The 100Hz here is bogus.  You could say, "expected to be run at 100Hz", but the code which actually implements that 100Hz is very much elsewhere.  There are many other modes which have similar comments in them - many wrong, all out-of-place :-)

Comment (rmackay9) [reply ArduCopter/stand_by.cpp]: yes, I agree we should remove these comments unless the function has some limitation that means it must run at only this speed (i.e. to stop devs from later modifying it).

File: ArduCopter/Copter.h (hunk 1 above)
Comment (lthall) [reply ArduCopter/Copter.h]: I completly agree!

File: ArduCopter/crash_check.cpp (hunk 2 above)
Comment (lthall) [reply ArduCopter/crash_check.cpp]: Cool, done.

File: ArduCopter/Copter.cpp (hunk 4 above)
Comment (lthall) [reply ArduCopter/Copter.cpp]: Cool.

File: ArduCopter/stand_by.cpp (hunk 5 above)
Comment (lthall) [reply ArduCopter/stand_by.cpp]: I have changed it to "Run standby functions at approximately 100 Hz to limit maximum variable build up" to make it more clear what the intended rate and purpose is. I hope that is more appropriate.

File: ArduCopter/land_detector.cpp (hunk 3 above)
Comment (lthall) [reply ArduCopter/land_detector.cpp]: Yep, I like it.

File: libraries/AC_AttitudeControl/AC_PosControl.cpp (hunk 6 above)
Comment (lthall) [reply libraries/AC_AttitudeControl/AC_PosControl.cpp]: No, this is correct. We ensure the aircraft can still fly in standby to prevent a crash if it is mistakenly given control of the aircraft when in standby.

File: libraries/AC_AttitudeControl/AC_PosControl.cpp (hunk 7 above)
Comment (lthall) [reply libraries/AC_AttitudeControl/AC_PosControl.cpp]: I have added comments for each line and corrected that one.

File: libraries/AC_WPNav/AC_Loiter.cpp (hunk 8 above)
Comment (lthall) [reply libraries/AC_WPNav/AC_Loiter.cpp]: Yes I will remove them.

Comment (lthall) [reply libraries/AC_WPNav/AC_Loiter.cpp]: Removed them

**Comment by lthall** [comment]
For some reason this failed:
CI_BUILD_TARGET="sitltest-rover sitltest-sub sitltest-balancebot"

**Comment by peterbarker** [comment]
On Tue, 8 Oct 2019, Leonard Hall wrote:

> For some reason this failed:
> CI_BUILD_TARGET="sitltest-rover sitltest-sub sitltest-balancebot"

Try rebasing on master and pushing up again, please.

File: libraries/RC_Channel/RC_Channel.h [hunk 9]
```diff
@@ -175,6 +175,7 @@ class RC_Channel {
         SURFACE_TRACKING =    75, // Surface tracking upwards or downwards
         KILL_IMU1 =          100, // disable first IMU (for IMU failure testing)
         KILL_IMU2 =          101, // disable second IMU (for IMU failure testing)
+        STAND_BY  =          102, // standy by functions
```
Comment (rmackay9) [inline libraries/RC_Channel/RC_Channel.h]: ah, the enum for STAND_BY should be 76.  I think the KILL_IMU stuff was purpose put high so it wouldn't accidentally be set by users (I'm not totally sure of the reason)

**Comment by rmackay9** [comment]
Merged, thanks!

**Comment by proficnc** [comment]
Awesome!!! A new era in ardupilot begins!

**Comment by Kiwa21** [comment]
Can we have more info on this ? @proficnc @lthall 
Should we expect a dual/tri cube architecture "soon" ?
Thanks

**Comment by GlidLov** [comment]
Hi all
I'm very interested in the StandBy mode, looking forward to create a fully rendundant FC. My question is about architecture and thus wiring. Culprit here is connection to ESCs: as today, escs (most of them, at least) have one single wire which has to be connected to an FC. Now, this leads to only one kind of system: a primary FC which has ESC connected, and a secondary FC which shares data with first one via mavlink (to ensure attitude solution is good, to be aware if the primary FC is working ok, etc). Then, when the second one is triggered, first one must release PWM outputs control to the second one, which will drive ESCs using part of first's HW. 
This solution would be inherently weak, because first FC must be still used somehow while second kicks in. A real rendundant one, would have a second set of PWM wires directly connected to the same ESC as first. Then also ESC must have a two-input discrimination system. 
So, do a working (on paper, at least) architecture has been shared already, on top of standby mode?
